src/compression.rs

```diff
@@ -64,6 +66,63 @@ const AVAILABLE_ENCODINGS: &[ContentCoding] = &[
     ContentCoding::ZSTD,
 ];
 
+/// Initializes dynamic compression.
+pub fn init(enabled: bool, handler_opts: &mut RequestHandlerOpts) {
+    handler_opts.compression = enabled;
+
+    const FORMATS: &[&str] = &[
+        #[cfg(any(feature = "compression", feature = "compression-deflate"))]
+        "deflate",
+        #[cfg(any(feature = "compression", feature = "compression-gzip"))]
+        "gzip",
+        #[cfg(any(feature = "compression", feature = "compression-brotli"))]
+        "brotli",
+        #[cfg(any(feature = "compression", feature = "compression-zstd"))]
+        "zstd",
+    ];
+    server_info!(
+        "auto compression: enabled={enabled}, formats={}",
+        FORMATS.join(",")
+    );
+}
+
+/// Post-processing to dynamically compress the response if necessary.
+pub(crate) fn post_process(
+    opts: &RequestHandlerOpts,
+    req: &Request<Body>,
+    mut resp: Response<Body>,
+) -> Result<Response<Body>, Error> {
+    if !opts.compression {
+        return Ok(resp);
+    }
+
+    let is_precompressed = resp.headers().get(CONTENT_ENCODING).is_some();
+    if is_precompressed {
+        return Ok(resp);
+    }
+
+    // Compression content encoding varies so use a `Vary` header
+    resp.headers_mut().append(
+        hyper::header::VARY,
+        HeaderValue::from_name(hyper::header::ACCEPT_ENCODING),
+    );
```

Comment:
I just noted that we are applying the header twice (also `src/compression_static.rs`) in case `--compression-static` is enabled, which is not by default but maybe we can add this header to some post-processing function to do it once if possible (it was the intention of the boolean flag after all). 

I also assume that this should be more clear once we tackle https://github.com/static-web-server/static-web-server/issues/362 later.

Reply:
Yes, we are applying it twice – but only in the comparably rare scenario where both static and dynamic compression are enabled. I don’t see an easy way to deal with that without intermingling `compression` and `compression_static` modules again when the goal here was actually making these modules self-contained. I only see a way that will introduce significant complexity.

There is no user-visible effect, the second call simply overwrites the first. And the performance impact should be negligible as well. So IMHO not worth addressing.